README.md
# nvidia-snatcher [![ci](https://github.com/jef/nvidia-snatcher/workflows/ci/badge.svg)](https://github.com/jef/nvidia-snatcher/actions?query=workflow%3Aci) [![discord](https://img.shields.io/discord/756303724095471617.svg?label=chat&logo=discord&logoColor=ffffff&color=7389D8)](https://discord.gg/Cyc7nrz)

[FAQ](#FAQ) | [Issues](https://github.com/jef/nvidia-snatcher/issues) | [Wiki](https://github.com/jef/nvidia-snatcher/wiki)

![nvidia-snatcher](media/screenshot.png)

The purpose of this bot is to get an Nvidia card. It tries multiple things to do that.

- Currently, `nvidia-snatcher` is not capable of purchasing a card for you
- Scrapes multiple websites for patterns of being stocked
    - API requests are a work in progress (very soon)
- Opens browser when stock is available
- Sends an email to you when stock is avaiable (must have Gmail)

<details>
<summary>What you may see if you're lucky</summary>

```sh
2020-09-18T07:06:28.535Z info :: ✖ [nvidia] nvidia founders edition is still out of stock
2020-09-18T07:06:31.241Z info :: ✖ [nvidia] nvidia founders edition is still out of stock
2020-09-18T07:06:34.212Z info :: ✖ [bestbuy] nvidia founder edition is still out of stock
2020-09-18T07:06:39.878Z info :: ✖ [bandh] gigabyte black is still out of stock
2020-09-18T07:06:43.236Z info :: ✖ [bestbuy] gigabyte black is still out of stock
2020-09-18T07:06:43.318Z info :: ↗ trying stores again
2020-09-18T07:06:43.318Z info :: 🚀🚀🚀 [nvidia] nvidia founders edition IN STOCK 🚀🚀🚀
2020-09-18T07:06:43.318Z info :: https://store.nvidia.com/store/nvidia/en_US/buy/productID.5438481700/clearCart.yes/nextPage.QuickBuyCartPage
```

</details>

| | Adorama | Amazon | Amazon (CA) | ASUS | B&H | Best Buy | Best Buy (CA) | EVGA | Micro Center | Newegg | Newegg (CA) | Nvidia | Office Depot | PNY | Zotac |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| **3070**| | | | | | | | | | | | | | `✔` | |
| **3080** | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` |
| **3090** | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | `✔` | | `✔` | `✔` | `✔` | | `✔` | `✔` |

## Installation and prerequisites

Linux, macOS, and Windows are all capable operating systems.

You do not need any computer skills, smarts, or anything of that nature. You are very capable as you have made it this far. Some basic understanding how a terminal, git, and or Node.js is a bonus, but that does not limit you to getting `nvidia-snatcher` running!

### Quick overview

#### Native installation

- [Node.js 14](https://nodejs.org/en/)
- [git](https://git-scm.com/)
- Clone this project `git clone https://github.com/jef/nvidia-snatcher.git`
- Run `npm install`
- Copy `.env.example` to a new file `.env` and edit the `.env` file to your liking using your [favorite text editor](https://code.visualstudio.com/)
    - More on this in [customization](#Customization)
- Run `npm run start` to start

At any point you want the program to stop, use <kbd>Ctrl</kbd> + <kbd>C</kbd>.

> :point_right: Please visit the [wiki](https://github.com/jef/nvidia-snatcher/wiki) if you need more help with installation.

#### Docker image (To run inside containers)

Available via GitHub Container Registry.

| Tag | Note |
|:---:|---|
| `latest` | Latest stable build |
| `nightly` | Latest HEAD build, could be unstable |

Use `docker run --cap-add=SYS_ADMIN -it --rm --env-file ./.env ghcr.io/jef/nvidia-snatcher:nightly` to run.

### Developer notes

The command `npm run start:dev` can be used instead of `npm run start` to automatically restart the project when filesystem changes are detected in the `src/` folder or `.env` file.

### Customization

To customize `nvidia-snatcher`, make a copy of `.env-example` as `.env` and make any changes to your liking. _All environment variables are **optional**._

Here is a list of variables that you can use to customize your newly copied `.env` file:

| Environment variable | Description | Notes |
|:---:|---|---|
| `BROWSER_TRUSTED` | Skip Chromium Sandbox | Useful for containerized environments, default: `false` |
| `COUNTRY` | [Supported country](#supported-countries) you want to be scraped | Currently only used by Nvidia, default: `usa` |
| `DESKTOP_NOTIFICATIONS` | Display desktop notifications using [node-notifier](https://www.npmjs.com/package/node-notifier) | Default: `false` |
| `DISCORD_NOTIFY_GROUP` | Discord group you would like to notify | Can be comma separated, use role ID, E.g.: `<@2834729847239842>` |
| `DISCORD_WEB_HOOK` | Discord Web Hook URL | Can be comma separated, use whole webhook URL |
| `EMAIL_PASSWORD` | Gmail password | See below if you have MFA |
| `EMAIL_TO` | Destination Email | Defaults to username if not set. Can be comma separated |
| `EMAIL_USERNAME` | Gmail address | E.g.: `jensen.robbed.us@gmail.com` |
| `HEADLESS` | Puppeteer to run headless or not | Debugging related, default: `true` |
| `IN_STOCK_WAIT_TIME` | Time to wait between requests to the same link if it has that card in stock | In seconds, default: `0` |
| `LOG_LEVEL` | [Logging levels](https://github.com/winstonjs/winston#logging-levels) | Debugging related, default: `info` |
| `LOW_BANDWIDTH` | Blocks images/fonts to reduce traffic | Disables ad blocker, default: `false` |
| `MAX_PRICE_SERIES_3070` | Maximum price allowed for a match, applies 3070 series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MAX_PRICE_SERIES_3080` | Maximum price allowed for a match, applies 3080 series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MAX_PRICE_SERIES_3090` | Maximum price allowed for a match, applies 3090 series cards (does not apply to these sites: Nvidia, Asus, EVGA) | Default: leave empty for no limit, otherwise enter a price (enter whole dollar amounts only, avoid use of: dollar symbols, commas, and periods.) e.g.: `1234` - Cards above `1234` will be skipped. |
| `MICROCENTER_LOCATION` | Specific MicroCenter location(s) to search | Comma separated, e.g.: `marietta,duluth`, default: `web` |
| `NVIDIA_ADD_TO_CART_ATTEMPTS` | The maximum number of times the `nvidia-api` add to cart feature will be attempted before failing | Default: `10` |
| `NVIDIA_SESSION_TTL` | The time in milliseconds to keep the cart active while using `nvidia-api` | Default: `60000` |
| `OPEN_BROWSER` | Toggle for whether or not the browser should open when item is found | Default: `true` |
| `PAGE_BACKOFF_MIN` | Minimum backoff time between retrying requests for the same store when a forbidden response is received | Default: `10000` |
| `PAGE_BACKOFF_MAX` | Maximum backoff time between retrying requests for the same store when a forbidden response is received | Default: `3600000` |
| `PAGE_SLEEP_MIN` | Minimum sleep time between queries of the same product page | In milliseconds, default: `5000` |
| `PAGE_SLEEP_MAX` | Maximum sleep time between queries of the same product page | In milliseconds, default: `10000` |
| `PAGE_TIMEOUT` | Navigation Timeout in milliseconds | `0` for infinite, default: `30000` |
| `PHONE_CARRIER` | [Supported carriers](#supported-carriers) for SMS | Email configuration required |
| `PHONE_NUMBER` | 10 digit phone number | E.g.: `1234567890`, email configuration required |
| `PLAY_SOUND` | Play this sound notification if a card is found | Relative path accepted, valid formats: wav, mp3, flac, E.g.: `path/to/notification.wav`, [free sounds available](https://notificationsounds.com/) |
| `PROXY_ADDRESS` | IP Address or fqdn of proxy server |
| `PROXY_PORT` | TCP Port number on which the proxy is listening for connections | Default: `80` |
| `PUSHBULLET` | PushBullet API key | Generate at https://www.pushbullet.com/#settings/account | |
| `PUSHOVER_TOKEN` | Pushover access token | Generate at https://pushover.net/apps/build | |
| `PUSHOVER_USER` | Pushover username | |
| `PUSHOVER_PRIORITY` | Pushover message priority |
| `SCREENSHOT` | Capture screenshot of page if a card is found | Default: `true` |
| `SHOW_ONLY_BRANDS` | Filter to show specified brands | Comma separated, e.g.: `evga,zotac` |
| `SHOW_ONLY_MODELS` | Filter to show specified models | Comma separated, e.g.: `founders edition,rog strix` |
| `SHOW_ONLY_SERIES` | Filter to show specified series | Comma separated, e.g.: `3080` |
| `SLACK_CHANNEL` | Slack channel for posting | E.g.: `update`, no need for `#` |
| `SLACK_TOKEN` | Slack API token | |
| `SMTP_ADDRESS` | IP Address or fqdn of smtp server |
| `SMTP_PORT` | TCP Port number on which the smtp server is listening for connections | Default: `25` |
| `STORES` | [Supported stores](#supported-stores) you want to be scraped | Comma separated, default: `nvidia` |
| `SCREENSHOT` | Capture screenshot of page if a card is found | Default: `true` |
| `TELEGRAM_ACCESS_TOKEN` | Telegram access token | |
| `TELEGRAM_CHAT_ID` | Telegram chat ID | |
| `TWILIO_ACCOUNT_SID` | Twilio Account SID | Can be found on twilio.com/console |
| `TWILIO_AUTH_TOKEN` | Twilio Auth Token | Can be found on twilio.com/console |
| `TWILIO_FROM_NUMBER` | Twilio provided phone number to send messages from | Include country code e.g +4401234567890 |
| `TWILIO_TO_NUMBER` | Mobile number to send SMS to | Include country code e.g +4401234567890 |
| `TWITTER_ACCESS_TOKEN_KEY` | Twitter Token Key | |
| `TWITTER_ACCESS_TOKEN_SECRET` | Twitter Token Secret | |
| `TWITTER_CONSUMER_KEY` | Twitter Consumer Key | Generate all Twitter keys at: https://developer.twitter.com/ |
| `TWITTER_CONSUMER_SECRET` | Twitter Consumer Secret | |
| `TWITTER_TWEET_TAGS` | Optional list of hashtags to append to the tweet message | E.g.: `#nvidia #nvidiastock` |
| `USER_AGENT` | Custom User-Agent header for HTTP requests | Default: `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36` |

> :point_right: If you have multi-factor authentication (MFA), you will need to create an [app password](https://myaccount.google.com/apppasswords) and use this instead of your Gmail password.

> :point_right: You can find your computer's user agent by [searching google for "my user agent"](http://google.com/search?q=my+user+agent)

> :point_right: You can test your notification configuration by running `npm run test:notification`.

> :point_right: Data usage is [known to be high](https://github.com/jef/nvidia-snatcher/issues?q=is%3Aissue+sort%3Aupdated-desc+bandwidth). This is expected as the program scrapes many websites in parallel 24/7. To help reduce this, use `LOW_BANDWIDTH="true"`. We are looking into other solutions as well, but is low priority.

#### Supported stores

| Stores | Environment variable |
|:---:|:---:|
| Adorama | `adorama`|
| Alternate (DE) | `alternate`|
| Amazon | `amazon`|
| Amazon (CA) | `amazon-ca`|
| Amazon (DE) | `amazon-de`|
| Amazon (NL) | `amazon-nl`|
| Amazon (UK) | `amazon-uk`|
| Aria PC | `aria`|
| ASUS | `asus` |
| B&H | `bandh`|
| Best Buy | `bestbuy`|
| Best Buy (CA) | `bestbuy-ca`|
| Box | `box`|
| Caseking | `caseking`|
| CCL | `ccl`|
| Coolblue | `coolblue`|
| Currys | `currys`|
| eBuyer | `ebuyer`|
| EVGA | `evga`|
| EVGA (EU) | `evga-eu`|
| Kabum (BR) | `kabum`|
| Gamestop | `gamestop`|
| Mediamarkt (DE) | `mediamarkt`|
| Micro Center | `microcenter`|
| Newegg | `newegg`|
| Newegg (CA) | `newegg-ca`|
| Notebooksbilliger (DE) |`notebooksbilliger`|
| Novatech | `novatech`|
| Nvidia | `nvidia`|
| Nvidia (API) | `nvidia-api`|
| Office Depot | `officedepot`|
| Pichau (BR) | `pichau`|
| PNY | `pny`|
| Terabyte (BR) | `terabyte`|
| Overclockers | `overclockers`|
| PNY | `pny`|
| Proshop (DE) | `proshop-de`|
| Proshop (DK) | `proshop-dk`|
| Scan | `scan`|
| Very | `very`|
| Zotac | `zotac`|

<details>
<summary>Micro Center stores</summary>

> :point_right: Before using `web`, please review [this issue comment](https://github.com/jef/nvidia-snatcher/issues/442#issuecomment-703297393).

| Store name |
|:---:|
| `brooklyn` |
| `brentwood` |
| `cambridge` |
| `chicago` |
| `columbus` |
| `dallas` |
| `denver` |
| `duluth` |
| `fairfax` |
| `flushing` |
| `houston` |
| `madison-heights` |
| `marietta` |
| `mayfield-heights` |
| `north-jersey` |
| `overland-park` |
| `parkville` |
| `rockville` |
| `sharonville` |
| `st-davids` |
| `st-louis-park` |
| `tustin` |
| `westbury` |
| `westmont` |
| `yonkers` |

</details>

#### Supported Brands and Models

| Brand | Model |
|:---:|---|
| `asus` | `rog strix`, `rog strix oc`, `strix`, `tuf`, `tuf oc` |
| `evga` | `ftw3`, `ftw3 ultra`, `xc3 black`, `xc3`, `xc3 ultra` |
| `gigabyte` | `aorus master`, `eagle`, `eagle oc`, `gaming`, `gaming oc`, `vision` |
| `msi` | `gaming x trio`, `ventus 3x`, `ventus 3x oc` |
| `nvidia` | `founders edition` |
| `pny` | `dual fan`, `xlr8`, `xlr8 rgb` |
| `zotac` | `trinity`, `trinity oc` |

#### Supported carriers

| Carrier | Environment variable | Notes |
|:---:|:---:|:---:|
| AT&T | `att` | |
| AT&T Prepaid | `attgo` | |
| Bell | `bell` | |
| Fido | `fido` | |
| Google | `google`| |
| Koodo | `koodo` | |
| Mint | `mint`| |
| Rogers | `rogers` | |
| Sprint | `sprint`| |
| Telus | `telus`| |
| T-Mobile | `tmobile`| |
| Verizon | `verizon`| Works with Visible |
| Virgin | `virgin`| |
| Virgin (CA) | `virgin-ca`| |

#### Supported countries

`COUNTRY` is only used by the `nvidia` and `nvidia-api` stores.

| Country | 3080 FE | 3090 FE | Test Card | Notes |
|:---:|:---:|:---:|:---:|:---:|
| austria | `✔` | `✔` | `✔` | |
| belgium | `✔` | `✔` | `✔` | |
| canada | `✔` | `✔` | `✔` | |
| czechia | `✔` | `✔` | `✔` | |
| denmark | `✔` | | `✔` | Missing RTX 3090 |
| finland | `✔` | | `✔` | Missing RTX 3090 |
| france | `✔` | `✔` | `✔` | |
| germany | `✔` | `✔` | `✔` | |
| great_britain | `✔` | `✔` | `✔` | |
| ireland | `✔` | `✔` | `✔` | |
| italy | `✔` | `✔` | `✔` | |
| luxembourg | `✔` | `✔` | `✔` | |
| netherlands | `✔` | `✔` | `✔` | |
| norway | `✔` | `✔` | `✔` | |
| poland | `✔` | `✔` | `✔` | |
| portugal | `✔` | | | RTX 3080 only |
| spain | `✔` | `✔` | `✔` | |
| sweden | `✔` | `✔` | `✔` | |
| usa | `✔` | `✔` | `✔` | |

## FAQ

**Q: What's Node.js and how do I install it?** Visit [their website](https://nodejs.org/en/) and download and install it. Very straight forward. Otherwise, Google more information related to your system needs.

**Q: Will this harm my computer?** No.

**Q: Have you gotten a card yet?** YES! :tada: :rocket:

<details>
<summary>Screenshot</summary>

![screenshot](https://i.imgur.com/59CRzGq.png)

</details>

**Q: Will I get banned from of the stores?** Perhaps, but getting a card is a nice outcome.

**Q: I got a problem and need help!** Join the [Discord](https://discord.gg/Cyc7nrz) or file an [issue](https://github.com/jef/nvidia-snatcher/issues/new/choose), I'll do my best to get to you. I work a full time job and this is only a hobby of mine.

**Q: How do I get the latest code?** Take look at this [wiki page](https://github.com/jef/nvidia-snatcher/wiki/Troubleshoot:-General:-Getting-the-latest-code)

**Q: Why don't my notifications work?** There are probably an [issue](https://github.com/jef/nvidia-snatcher/issues?q=is%3Aissue+sort%3Aupdated-desc+sound+is%3Aclosed) [that] has [already](https://github.com/jef/nvidia-snatcher/issues/182) [been](https://github.com/jef/nvidia-snatcher/issues/116) [resolved](https://github.com/jef/nvidia-snatcher/issues/155)

**Q: I'd love to contribute, how do I do that?** Make a [pull request](https://github.com/jef/nvidia-snatcher/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc)! All contributions are welcome.

**Q: Why do I have to download all this stuff just to get this bot working?** Well, I would rather you didn't either. See [#11](https://github.com/jef/nvidia-snatcher/issues/11).

### Acknowledgements

Thanks to the great contributors that make this project possible

Special shout to initial developers:

- [@andirew](https://github.com/andirew)
- [@fuckingrobot](https://github.com/fuckingrobot)
- [@ioncaza](https://github.com/IonCaza)
- [@malbert69](https://github.com/malbert69)


CHANGELOG.md
# Changelog

## [1.6.0](https://www.github.com/danmaxis/nvidia-snatcher/compare/v1.5.0...v1.6.0) (2020-10-15)


### Features

* add `meta` to logger ([#437](https://www.github.com/danmaxis/nvidia-snatcher/issues/437)) ([ae6bc86](https://www.github.com/danmaxis/nvidia-snatcher/commit/ae6bc86bcb75c154a2a68adda324f34f18281700))
* add asus-de store and asus 3080 tuf/oc to amazon-de ([#436](https://www.github.com/danmaxis/nvidia-snatcher/issues/436)) ([b4d8733](https://www.github.com/danmaxis/nvidia-snatcher/commit/b4d8733d1823e2fa5d3a1b80d2d4218bfa737cd7))
* add caseking and proshop-de store ([#521](https://www.github.com/danmaxis/nvidia-snatcher/issues/521)) ([1aaa102](https://www.github.com/danmaxis/nvidia-snatcher/commit/1aaa102c06281b60da593d04ced01f8b20a37a3b))
* add coolblue store ([#482](https://www.github.com/danmaxis/nvidia-snatcher/issues/482)) ([5c61333](https://www.github.com/danmaxis/nvidia-snatcher/commit/5c61333df00c0c30f04e4aced01a498c5816ad58))
* add german stores ([#462](https://www.github.com/danmaxis/nvidia-snatcher/issues/462)) ([85a07dc](https://www.github.com/danmaxis/nvidia-snatcher/commit/85a07dc0dc6a8dfe2aba1d4ea3e4cde6ec083086))
* add link series to error message for better context ([#265](https://www.github.com/danmaxis/nvidia-snatcher/issues/265)) ([332b4a8](https://www.github.com/danmaxis/nvidia-snatcher/commit/332b4a8246320e458729c0b58c310d290df12530))
* add proshop-dk ([#524](https://www.github.com/danmaxis/nvidia-snatcher/issues/524)) ([dc2fcf5](https://www.github.com/danmaxis/nvidia-snatcher/commit/dc2fcf5ab4dba8be459bd5aa6a41e3e097205d45))
* add rog strix oc to amazon-de ([#471](https://www.github.com/danmaxis/nvidia-snatcher/issues/471)) ([e5909c5](https://www.github.com/danmaxis/nvidia-snatcher/commit/e5909c532da1c8b731e9835bffd8f31b76550d4d))
* add support for specifying smtp server ([#458](https://www.github.com/danmaxis/nvidia-snatcher/issues/458)) ([160ae37](https://www.github.com/danmaxis/nvidia-snatcher/commit/160ae37d7bee4f0323cb939566d586f1300aece6))
* allow multiple microcenter locations ([#487](https://www.github.com/danmaxis/nvidia-snatcher/issues/487)) ([2b0eab6](https://www.github.com/danmaxis/nvidia-snatcher/commit/2b0eab656d81314ce7d05064662ec138407550c3))
* clear cookies and cache ([#515](https://www.github.com/danmaxis/nvidia-snatcher/issues/515)) ([1f89945](https://www.github.com/danmaxis/nvidia-snatcher/commit/1f89945c5746befb2131381a358efdf8ac530e0c)), closes [#417](https://www.github.com/danmaxis/nvidia-snatcher/issues/417)
* **docker:** add docker and publish images to ghcr ([#411](https://www.github.com/danmaxis/nvidia-snatcher/issues/411)) ([c857985](https://www.github.com/danmaxis/nvidia-snatcher/commit/c857985a6d3736287976caf5b173f19046306465))
* **notification:** add `to` field for email ([#327](https://www.github.com/danmaxis/nvidia-snatcher/issues/327)) ([8828dd1](https://www.github.com/danmaxis/nvidia-snatcher/commit/8828dd15cd08959cd434bd256e8eac474dd17c49))
* **notification:** add at&t prepaid carrier ([#425](https://www.github.com/danmaxis/nvidia-snatcher/issues/425)) ([eb9d082](https://www.github.com/danmaxis/nvidia-snatcher/commit/eb9d082d178a42e35789ba822cfae7b35c0413c1))
* **notification:** add pushover priority setting ([#186](https://www.github.com/danmaxis/nvidia-snatcher/issues/186)) ([f277172](https://www.github.com/danmaxis/nvidia-snatcher/commit/f2771721914a20619833df8ccb2ac44298687b4d))
* **notification:** add Twillio notification provider ([#344](https://www.github.com/danmaxis/nvidia-snatcher/issues/344)) ([f2f8d81](https://www.github.com/danmaxis/nvidia-snatcher/commit/f2f8d81498d1acfb9359f4a690962042ec20d166))
* **store:** add `pny` ([#295](https://www.github.com/danmaxis/nvidia-snatcher/issues/295)) ([f6760d3](https://www.github.com/danmaxis/nvidia-snatcher/commit/f6760d3c65d60eae9e4b1e0fdba34e814f446a4c))
* **store:** add 3090s to amazon-ca ([#274](https://www.github.com/danmaxis/nvidia-snatcher/issues/274)) ([e992cf4](https://www.github.com/danmaxis/nvidia-snatcher/commit/e992cf4db85f045fc8d03c9b93286fb72bad1061))
* **store:** add additional cards to all stores ([#286](https://www.github.com/danmaxis/nvidia-snatcher/issues/286)) ([d25a643](https://www.github.com/danmaxis/nvidia-snatcher/commit/d25a643425020fa3f7cd48972360ede17501afeb))
* **store:** add amazon cards and `cartUrl`s ([#284](https://www.github.com/danmaxis/nvidia-snatcher/issues/284)) ([d69189f](https://www.github.com/danmaxis/nvidia-snatcher/commit/d69189f12c893fb6d88b198d802ff8e36a69bc88))
* **store:** add evga 3090 to newegg canada ([#396](https://www.github.com/danmaxis/nvidia-snatcher/issues/396)) ([313d176](https://www.github.com/danmaxis/nvidia-snatcher/commit/313d176848a25f183334db8926ac0ec445a2c481))
* **store:** add ftw3 ultra to microcenter ([#448](https://www.github.com/danmaxis/nvidia-snatcher/issues/448)) ([5109227](https://www.github.com/danmaxis/nvidia-snatcher/commit/5109227e8899f57b828b157309a76c397c872559))
* **store:** add rog-strix-3080 to newegg-ca ([#489](https://www.github.com/danmaxis/nvidia-snatcher/issues/489)) ([207bc41](https://www.github.com/danmaxis/nvidia-snatcher/commit/207bc411ed240cd9150e018b21e735ba08a4d134))
* add uk stores ([#455](https://www.github.com/danmaxis/nvidia-snatcher/issues/455)) ([b9b6b55](https://www.github.com/danmaxis/nvidia-snatcher/commit/b9b6b55c29d11f48b683816e5b8c1cab127ed5fd))
* configurable status code behaviours ([#340](https://www.github.com/danmaxis/nvidia-snatcher/issues/340)) ([3b7487e](https://www.github.com/danmaxis/nvidia-snatcher/commit/3b7487e97ac9d93344403f50153f2de6243b1f0d))
* max price filtering ([#383](https://www.github.com/danmaxis/nvidia-snatcher/issues/383)) ([fd294d2](https://www.github.com/danmaxis/nvidia-snatcher/commit/fd294d2baa06a1c0a68852497889a0412dea492e))
* max price per series ([#451](https://www.github.com/danmaxis/nvidia-snatcher/issues/451)) ([8adc07a](https://www.github.com/danmaxis/nvidia-snatcher/commit/8adc07a03e411dd536bebfdc7270db4bbf8ddb34))

src/store/model/index.ts
import {Adorama} from './adorama';
import {Alternate} from './alternate';
import {Amazon} from './amazon';
import {AmazonCa} from './amazon-ca';
import {AmazonDe} from './amazon-de';
import {AmazonNl} from './amazon-nl';
import {AmazonUk} from './amazon-uk';
import {Aria} from './aria';
import {Asus} from './asus';
import {AsusDe} from './asus-de';
import {BAndH} from './bandh';
import {BestBuy} from './bestbuy';
import {BestBuyCa} from './bestbuy-ca';
import {Box} from './box';
import {Caseking} from './caseking';
import {Ccl} from './ccl';
import {Coolblue} from './coolblue';
import {Currys} from './currys';
import {Ebuyer} from './ebuyer';
import {Evga} from './evga';
import {EvgaEu} from './evga-eu';
import {GKInfoStore} from './gkinfostore';
import {Gamestop} from './gamestop';
import {Kabum} from './kabum';
import {Mediamarkt} from './mediamarkt';
import {MicroCenter} from './microcenter';
import {Newegg} from './newegg';
import {NeweggCa} from './newegg-ca';
import {Notebooksbilliger} from './notebooksbilliger';
import {Novatech} from './novatech';
import {Nvidia} from './nvidia';
import {NvidiaApi} from './nvidia-api';
import {OfficeDepot} from './officedepot';
import {Overclockers} from './overclockers';
import {Pichau} from './pichau';
import {Pny} from './pny';
import {ProshopDE} from './proshop-de';
import {ProshopDK} from './proshop-dk';
import {Scan} from './scan';
import {Store} from './store';
import {Terabyte} from './terabyte';
import {Very} from './very';
import {Zotac} from './zotac';

import {config} from '../../config';
import {logger} from '../../logger';

const masterList = new Map([
	[Adorama.name, Adorama],
	[Alternate.name, Alternate],
	[Amazon.name, Amazon],
	[AmazonCa.name, AmazonCa],
	[AmazonDe.name, AmazonDe],
	[AmazonNl.name, AmazonNl],
	[AmazonUk.name, AmazonUk],
	[Aria.name, Aria],
	[Asus.name, Asus],
	[AsusDe.name, AsusDe],
	[BAndH.name, BAndH],
	[BestBuy.name, BestBuy],
	[BestBuyCa.name, BestBuyCa],
	[Box.name, Box],
	[Caseking.name, Caseking],
	[Ccl.name, Ccl],
	[Coolblue.name, Coolblue],
	[Currys.name, Currys],
	[Ebuyer.name, Ebuyer],
	[Evga.name, Evga],
	[EvgaEu.name, EvgaEu],
	[Kabum.name, Kabum],
	[Gamestop.name, Gamestop],
	[GKInfoStore.name, GKInfoStore],
	[Mediamarkt.name, Mediamarkt],
	[MicroCenter.name, MicroCenter],
	[Newegg.name, Newegg],
	[NeweggCa.name, NeweggCa],
	[Notebooksbilliger.name, Notebooksbilliger],
	[Novatech.name, Novatech],
	[Nvidia.name, Nvidia],
	[NvidiaApi.name, NvidiaApi],
	[OfficeDepot.name, OfficeDepot],
	[Overclockers.name, Overclockers],
	[Pichau.name, Pichau],
	[ProshopDE.name, ProshopDE],
	[ProshopDK.name, ProshopDK],
	[Pny.name, Pny],
	[Terabyte.name, Terabyte],
	[Scan.name, Scan],
	[Very.name, Very],
	[Zotac.name, Zotac]
]);

const list = new Map();

for (const name of config.store.stores) {
	if (masterList.has(name)) {
		list.set(name, masterList.get(name));
	} else {
		const logString = `No store named ${name}, skipping.`;
		logger.warn(logString);
	}
}

logger.info(`ℹ selected stores: ${Array.from(list.keys()).join(', ')}`);

if (config.store.showOnlyBrands.length > 0) {
	logger.info(`ℹ selected brands: ${config.store.showOnlyBrands.join(', ')}`);
}

if (config.store.showOnlyModels.length > 0) {
	logger.info(`ℹ selected models: ${config.store.showOnlyModels.join(', ')}`);
}

if (config.store.showOnlySeries.length > 0) {
	logger.info(`ℹ selected series: ${config.store.showOnlySeries.join(', ')}`);
}

export const Stores = Array.from(list.values()) as Store[];

export * from './store';


src/store/model/kabum.ts
import {Store} from './store';

export const Kabum: Store = {
	labels: {
		inStock: {
			container: '.botao-comprar',
			text: ['']
		}
	},
	links: [
		{
			brand: 'test:brand',
			model: 'test:model',
			series: 'test:series',
			url: 'https://www.kabum.com.br/produto/100169/kindle-novo-paperwhite-8gb-wi-fi-preto-ao0705'
		},
		{
			brand: 'asus',
			model: 'tuf',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/121138/placa-de-v-deo-asus-nvidia-geforce-rtx3080-10gb-gddr6-tuf-rtx3080-10g-gaming'
		},
		{
			brand: 'zotac',
			model: 'trinity',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/126895/placa-de-v-deo-zotac-nvidia-geforce-rtx-3080-10gb-gddr6x-zt-a30800d-10p'
		},
		{
			brand: 'msi',
			model: 'gaming-x-trio',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/127409/placa-de-v-deo-msi-nvidia-geforce-rtx-3080-gaming-x-trio-10g-gddr6x'
		},
		{
			brand: 'msi',
			model: 'ventus-3x',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/127410/placa-de-v-deo-msi-nvidia-geforce-rtx-3080-ventus-3x-10g-oc-gddr6x'
		},
		{
			brand: 'gainward',
			model: 'phoenix',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/128775/placa-de-v-deo-gainward-nvidia-geforce-nv-rtx3080-10gb-gddr6x-ned3080019ia-132ax'
		},
		{
			brand: 'evga',
			model: 'xc3',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/128051/placa-de-v-deo-evga-nvidia-geforce-rtx-3080-10gb-gddr6x-10g-p5-3883-kr'
		},
		{
			brand: 'gigabyte',
			model: 'eagle-oc',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/128078/placa-de-v-deo-gigabyte-nvidia-geforce-rtx-3080-10gb-gddr6x-gv-n3080eagle-oc-10gd'
		},
		{
			brand: 'msi',
			model: 'gaming-x-trio',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/127407/placa-de-v-deo-msi-nvidia-geforce-rtx-3090-gaming-x-trio-24g-gddr6x-'
		},
		{
			brand: 'msi',
			model: 'ventus-3x-oc',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/127408/placa-de-v-deo-msi-nvidia-geforce-rtx-3090-ventus-3x-24g-oc-gddr6x'
		},
		{
			brand: 'asus',
			model: 'tuf',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/127511/placa-de-v-deo-asus-nvidia-tuf-rtx-3090-24gb-gddr6x-90yv0fd1-m0nm00'
		},
		{
			brand: 'asus',
			model: 'tuf',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/121139/placa-de-v-deo-asus-nvidia-geforce-rtx3090-24gb-gddr6-tuf-rtx3090-24g-gaming'
		},
		{
			brand: 'asus',
			model: 'rog-strix',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/128026/placa-de-v-deo-asus-nvidia-geforce-rtx-3090-24gb-gddr6x-rog-strix-rtx3090-o24g-gaming-'
		},
		{
			brand: 'gainward',
			model: 'phoenix',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/128776/placa-de-v-deo-gainward-nvidia-geforce-nv-rtx3090-24gb-gddr6x-ned3090019sb-132bx'
		},
		{
			brand: 'evga',
			model: 'xc3',
			series: '3090',
			url: 'https://www.kabum.com.br/produto/128375/placa-de-v-deo-evga-nvidia-geforce-rtx-3090-24gb-gddr6x-24g-p5-3975-kr'
		},
		{
			brand: 'gigabyte',
			model: 'eagle-oc',
			series: '3080',
			url: 'https://www.kabum.com.br/produto/128078/placa-de-v-deo-gigabyte-nvidia-geforce-rtx-3080-10gb-gddr6x-gv-n3080eagle-oc-10gd'
		}
	],
	name: 'kabum'
};
